File: android/supercamera/src/main/java/com/dcamera/core/CameraPorts.kt
package com.dcamera.core

import com.dcamera.capture.CaptureController
import com.dcamera.capture.CaptureResult
import com.dcamera.render.PreviewFrame

interface PreviewSessionPort {
    fun setFrameConsumer(consumer: (PreviewFrame) -> Unit)
    fun startPreview(surfaceHandle: Long): Boolean
    fun stopPreview(): Boolean
}

interface PhotoCapturePort {
    fun capture(outputPath: String): CaptureResult
}

class NoOpPreviewSessionPort : PreviewSessionPort {
    override fun setFrameConsumer(consumer: (PreviewFrame) -> Unit) {}
    override fun startPreview(surfaceHandle: Long): Boolean = surfaceHandle != 0L
    override fun stopPreview(): Boolean = true
}

class CaptureControllerAdapter(
    private val controller: CaptureController = CaptureController(),
) : PhotoCapturePort {
    override fun capture(outputPath: String): CaptureResult = controller.takePhoto(outputPath)
}

File: ios/SuperCameraKit/Sources/SuperCameraKit/CameraPorts.swift
import Foundation

public protocol PreviewSessionPort {
    func setFrameConsumer(_ consumer: @escaping (PreviewFrame) -> Void)
    func startPreview(surfaceHandle: UInt64) -> Bool
    func stopPreview() -> Bool
}

public protocol PreviewBridgePort {
    func attach(surfaceHandle: UInt64) -> Bool
    func detach()
    func submitFrame(frame: PreviewFrame) -> Bool
}

public protocol PhotoCapturePort {
    func capture(outputPath: String) -> CaptureResult
}

public final class NoOpPreviewSessionPort: PreviewSessionPort {
    public init() {}

    public func setFrameConsumer(_ consumer: @escaping (PreviewFrame) -> Void) {}

    public func startPreview(surfaceHandle: UInt64) -> Bool { surfaceHandle != 0 }

    public func stopPreview() -> Bool { true }
}

public final class CaptureControllerAdapter: PhotoCapturePort {
    private let controller: CaptureController

    public init(controller: CaptureController = CaptureController()) {
        self.controller = controller
    }

    public func capture(outputPath: String) -> CaptureResult {
        controller.takePhoto(outputPath: outputPath)
    }
}
